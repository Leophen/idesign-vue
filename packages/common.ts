import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
