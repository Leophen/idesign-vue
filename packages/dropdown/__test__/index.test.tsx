import { vi } from 'vitest';
import { Dropdown } from '../index';
import { render } from '../../common'

const options = [
  {
    content: '操作一',
    value: 1
  },
  {
    content: '操作二',
    value: 2
  },
  {
    content: '操作三',
    value: 3
  }
]

describe('Dropdown 组件测试', () => {
  it('create', () => {
    const wrapper = render(
      <Dropdown options={options}>
        <button>更多</button>
      </Dropdown>
    );
    expect(wrapper.element).toMatchSnapshot();
  });

  it('onClick', () => {
    const clickFn = vi.fn();
    const wrapper = render(
      <Dropdown options={options} onClick={clickFn}>
        <button>更多</button>
      </Dropdown>
    );
    expect(wrapper.element).toMatchSnapshot();
  });

  it('onTrigger', () => {
    const triggerFn = vi.fn();
    const wrapper = render(
      <Dropdown options={options} onTrigger={triggerFn}>
        <button>更多</button>
      </Dropdown>
    );
    expect(wrapper.element).toMatchSnapshot();
  });
});
