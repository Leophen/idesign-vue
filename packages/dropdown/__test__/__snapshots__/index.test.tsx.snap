// Vitest Snapshot v1

exports[`Dropdown 组件测试 > create 1`] = ``;

exports[`Dropdown 组件测试 > onClick 1`] = ``;

exports[`Dropdown 组件测试 > onTrigger 1`] = ``;
